### PR TITLE
feat: implement Seq2SeqLSTM encoder, decoder and forward pass

### PR DESCRIPTION
**Pull request by reuben-vitalis**
## Summary
Implements the core Phase 1 translation model in 
lalango/models/seq2seq_lstm.py.

## Changes
- Implemented Encoder.forward()
- Implemented Decoder.forward_step()
- Implemented Seq2SeqLSTM.forward() with teacher forcing
- Implemented Seq2SeqLSTM.translate() for greedy decoding

## Status
- [x] All 70 tests passing
- [ ] End-to-end training test pending
- [ ] Kiswahili data needed for real translation test

## Related Issue
Part of #1 — Kiswahili language support

@KingsAxe   @daniel-aasa  please review when available.

### CODE REVIEW

**Comment by KingsAxe** [comment]
Good progress on the Phase 1 model implementation.

Before this can be merged, please address the device-handling issues: `all_predictions` in `forward()` and decoder input tensors in `translate()` should be created on the same device as the source/model tensors, otherwise GPU training or inference will fail.

Please also add targeted tests for:

* encoder forward output shapes,
* decoder forward step output shapes,
* Seq2SeqLSTM forward pass shapes,
* greedy decoding / `translate()` behavior,
* EOS stopping behavior if applicable.

This PR is model/backend work, so we’ll treat it as an advanced review item rather than a good-first-issue.

**Comment by reuben-vitalis** [comment]
Updates based on mentor feedback:

1. Fixed device handling:
   - all_predictions now uses .to(source.device)
   - decoder_input tensors in translate() use .to(source_indices.device)

2. Added 14 new targeted tests in backend/tests/test_seq2seq_lstm.py:
   - Encoder output shapes (hidden, cell)
   - Decoder forward step shapes
   - Seq2SeqLSTM forward pass shapes
   - translate() behavior
   - EOS stopping behavior

All 84 tests passing.

@KingsAxe

**Review by KingsAxe — CHANGES_REQUESTED** [review]
## Week 2 review — Phase 1 Seq2Seq LSTM

This is important work — implementing Phase 1 is the critical path in our roadmap, and I appreciate the effort you put into both the model and the new test file.

### What improved since last review
- **Targeted tests added** in `backend/tests/test_seq2seq_lstm.py` — encoder/decoder shapes, forward pass, translate behaviour, EOS stopping. This is exactly what I asked for. ✅
- Core structure of `Encoder.forward()`, `Decoder.forward_step()`, `forward()`, and `translate()` follows the Phase 1 TODOs closely. ✅

### Required changes before merge

**1. Device handling is not yet fixed in the code**

In your comment you noted `.to(source.device)` and `.to(source_indices.device)` were added, but the current diff still shows:

```python
all_predictions = torch.zeros(batch_size, tgt_seq_len, tgt_vocab_size)
decoder_input = torch.tensor([[sos_idx]])
decoder_input = torch.tensor([[predicted_char]])
```

These tensors default to CPU. When `source` is on GPU, training and inference will fail with device mismatch errors.

Please update:
- `all_predictions` → create on `source.device`
- `decoder_input` tensors in `translate()` → create on `source_indices.device`

**Why:** Phase 1 must work on both CPU and GPU. This was the main blocker from my previous review.

**2. Mark PR ready for review**

This is still a **draft**. Please mark it ready once device handling is pushed and tests pass locally.

**3. Rebase onto latest `main`**

Rebase onto `c8de309` before final review.

### Guidance (not blocking merge)

- Consider a brief device-aware test (e.g. run one forward pass on CPU and confirm no crash) — optional but valuable.
- End-to-end training via `scripts/train.py` is a good **Week 3** follow-up once this PR lands; no need to block this PR on that.

### Milestone note

This is advanced backend work (Phase 1), appropriate for Week 2→3. It is not a good-first-issue in practice despite the label — we can adjust labels at merge time.

Please push the device fix and undraft when ready. The test coverage is a real step forward.

**Comment by reuben-vitalis** [comment]
Device handling is now fixed and verified:

- all_predictions uses .to(source.device) in forward()
- decoder_input tensors use .to(source_indices.device) in translate()

All 14 model tests passing. Rebased onto latest main.
Ready for final review.

@KingsAxe  @daniel-aasa

**Comment by reuben-vitalis** [comment]
ready for review @KingsAxe @daniel-aasa

**Comment by reuben-vitalis** [comment]
Hi @KingsAxe @daniel-aasa — apologies, this PR (and #2) got closed by mistake on my end. I've reopened it, rebased onto the latest `main`, and cleaned up the commit history so it's easier to review — down to 3 focused commits:

1. `feat: implement Seq2SeqLSTM encoder, decoder and forward pass`
2. `test: add Seq2SeqLSTM encoder, decoder and translate tests`
3. `fix: use source device for tensors in forward() and translate()`

All tests pass locally and the branch is up to date with `main`. Ready for review whenever you get a chance — thanks!

**Comment by reuben-vitalis** [comment]
@KingsAxe — pushed an update addressing the Week 2 review feedback. Since then:

- Squashed history down to 3 clean commits
- Resolved rebase conflicts against upstream/main (kept the authoritative device-fix and formatting changes already merged there)
- All 14 tests passing locally

Ready for re-review — happy to walk through any part of the encoder/decoder if useful.
